qemu/tests/timerdevice_check_ntp_offset.py

```diff
@@ -0,0 +1,96 @@
+
+    def scp_file(file_dir, file_name, file_dst_path):
+        file_src_path = os.path.join(data_dir.get_deps_dir(), file_dir, file_name)
+        vm.copy_files_to(file_src_path, file_dst_path)
+
+    def clean_tmp_file():
+        if not session.cmd_status("dir %s" % ntp_dst_path):
+            session.cmd("rd /s /q %s" % ntp_dst_path)
+        ntp_install_path = params["ntp_install_path"]
+        ntp_uninstall_cmd = params["ntp_uninstall_cmd"]
+        if not session.cmd_status("dir %s" % ntp_install_path):
+            session.cmd(ntp_uninstall_cmd)
+        diskspd_check_cmd = params["diskspd_check_cmd"]
+        diskspd_end_cmd = params["diskspd_end_cmd"]
+        if not session.cmd_status("dir %s" % (diskspd_dst_path + diskspd_name)):
+            if not session.cmd_status(diskspd_check_cmd):
+                session.cmd(diskspd_end_cmd)
+            session.cmd("del %s" % (diskspd_dst_path + diskspd_name))
+    ntp_cmd = params["ntp_cmd"]
+    error_context.context("Sync host system time with ntpserver", logging.info)
+    process.system(ntp_cmd, shell=True)
+
+    vm = env.get_vm(params["main_vm"])
+    session = vm.wait_for_login()
+
+    error_context.context("Install diskspd tool on guest", logging.info)
+    diskspd_dir = params["diskspd_dir"]
+    diskspd_name = params["diskspd_name"]
+    diskspd_dst_path = params["diskspd_dst_path"]
+    scp_file(diskspd_dir, diskspd_name, diskspd_dst_path)
+
+    error_context.context("Install Meinberg NTP on guest", logging.info)
+    ntp_dir = params["ntp_dir"]
+    ntp_name = params["ntp_name"]
+    ntp_unattend_file = params["ntp_unattend_file"]
+    ntp_conf = params["ntp_conf"]
+    ntp_dst_path = params["ntp_dst_path"]
+    install_ntp_cmd = params["install_ntp_cmd"]
+    session.cmd("md %s && cd %s" % (ntp_dst_path, ntp_dst_path))
+    scp_file(ntp_dir, ntp_name, ntp_dst_path)
+    scp_file(ntp_dir, ntp_unattend_file, ntp_dst_path)
+    scp_file(ntp_dir, ntp_conf, ntp_dst_path)
+    ntp_unattend_file_path = os.path.join(ntp_dst_path, ntp_unattend_file)
+    session.cmd(install_ntp_cmd % (ntp_name, ntp_unattend_file))
+
+    error_context.context("Run diskspd on guest", logging.info)
+    diskspd_run_cmd = params["diskspd_run_cmd"]
+    session.cmd("cd %s" % diskspd_dst_path)
+    session.cmd(diskspd_run_cmd)
+
+    error_context.context("Play a video on guest", logging.info)
+    sub_test = params["sub_test"]
+    utils_test.run_virt_sub_test(test, params, env, sub_test)
+
+    error_context.context("Check offset of ntp", logging.info)
+    check_offset_cmd = params["check_offset_cmd"]
+    sleep_time = params["sleep_time"]
+    nums = int(params["nums"])
+    try:
+        while nums > 0:
+            time.sleep(int(sleep_time))
+            ntp_offset = session.cmd_output(check_offset_cmd)
+            ntp_offset = float(ntp_offset.strip().split("\n")[-1].split()[-2].
+                               strip('-+'))
+            if ntp_offset > 100:
+                clean_tmp_file()
```

Comment:
Maybe the `clean_tmp_file()` is redundant since if an exception occurs in any of the clauses and is not handled, the exception is temporarily saved. The finally clause is executed.

Reply:
Agree,  this if branch means `test.fail`, and it can be caught by `finally`.